x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/LifecyclePolicy.java

```diff
@@ -245,6 +245,74 @@ public boolean isActionSafe(StepKey stepKey) {
         }
     }
 
+    public StepKey getNextValidStep(StepKey stepKey) {
+        Phase phase = phases.get(stepKey.getPhase());
+        if (phase == null) {
+            // Phase doesn't exist so find the after step for the previous
+            // available phase
+            return getStepBeforePhase(stepKey.getPhase());
+        } else {
+            // Phase exists so check if the action exists
+            LifecycleAction action = phase.getActions().get(stepKey.getAction());
+            if (action == null) {
+                // if action doesn't exist find the first step in the next
+                // available action
+                return getFirstStepInNextAction(stepKey.getAction(), phase);
+            } else {
+                // if the action exists check if the step itself exists
+                if (action.toStepKeys(phase.getName()).contains(stepKey)) {
+                    // stepKey is valid still so return it
+                    return stepKey;
+                } else {
+                    // stepKey no longer exists in the action so we need to move
+                    // to the first step in the next action since skipping steps
+                    // in an action is not safe
+                    return getFirstStepInNextAction(stepKey.getAction(), phase);
+                }
+            }
+        }
+    }
+    
+    private StepKey getNextAfterStep(String currentPhaseName) {
+        String nextPhaseName = type.getNextPhaseName(currentPhaseName, phases);
+        if (nextPhaseName == null) {
+            // We don't have a next phase after this one so the next step is the TerminalPolicyStep
+            return TerminalPolicyStep.KEY;
+        } else {
+            return new StepKey(currentPhaseName, PhaseAfterStep.NAME, PhaseAfterStep.NAME);
+        }
+    }
+
+    private StepKey getStepBeforePhase(String currentPhaseName) {
+        String nextPhaseName = type.getNextPhaseName(currentPhaseName, phases);
+        if (nextPhaseName == null) {
+            // We don't have a next phase after this one so the next step is the
+            // TerminalPolicyStep
+            return TerminalPolicyStep.KEY;
```

Comment:
I'm not sure I follow why the step before a phase would be the terminal step?

Reply:
oh, I think I get it. you are looking for the after-step for the next phase, but there is no next phase, so we go straight to terminalpolicystep. If this is the case, some extra docs would help make that clearer

Reply:
Yes you are correct, I'll try to clarify it better int eh comment